[migrate] upgrade to ESLint 9, Node.js 20 & other latest Upstream packages/actions

Soecka opened the pull request:
[![PR-16](https://badgen.net/badge/GitHub.dev/PR-16/blue?icon=visualstudio)](https://github.dev/idea2app/React-MobX-Bootstrap-ts/pull/16) [![PR-16](https://badgen.net/badge/GitHub%20codespaces/PR-16/black?icon=github)](https://codespaces.new/idea2app/React-MobX-Bootstrap-ts/pull/16) [![PR-16](https://badgen.net/badge/GitPod.io/PR-16/orange?icon=git)](https://gitpod.io/?autostart=true#https://github.com/idea2app/React-MobX-Bootstrap-ts/pull/16) [<img width="16" alt="Powered by Pull Request Badge" src="https://user-images.githubusercontent.com/1393946/111216524-d2bb8e00-85d4-11eb-821b-ed4c00989c02.png">](https://pullrequestbadge.com/?utm_medium=github&utm_source=idea2app&utm_campaign=badge_info)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

[add] eslint, lint-staged
[update] dependencies

Soecka committed:
[add] eslint, lint-staged
[update] dependencies

TechQuery commented on package.json:
```diff
@@ -66,7 +75,10 @@
     },
     "lint-staged": {
-        "*.{html,md,less,json,yml,js,ts,tsx}": "prettier --write"
+        "*.{html,md,less,json,yml,js,ts,tsx}": [
+            "eslint --fix",
+            "prettier --write"
+        ]
```
先格式化再检查代码规范会比较好

Soecka replied:
格式不对会影响 eslint 规则通过？

TechQuery replied:
> 格式不对会影响 eslint 规则通过？

有少数规则会，因为代码越格式规范出问题的概率越小。

Soecka replied:
官方文档推荐写的，一个是代码质量规则，一个是格式化规则，它们的交集也用插件 disable 了，很难想象格式会影响代码质量检查
https://github.com/lint-staged/lint-staged#task-concurrency

Soecka replied:
按照步骤，只要 `eslint --fix` 出问题都会再过一遍，用 prettier 收尾是正常流程？

TechQuery commented on package.json:
```diff
@@ -5,52 +5,61 @@
-        "@types/lodash": "^4.17.5",
-        "@types/node": "^18.19.39",
+        "@types/lodash": "^4.17.7",
+        "@types/node": "^20.16.2",
```
其它依赖 Node.js 20 的 GitHub actions 也都升级一下，参见：idea2app/Next-Bootstrap-ts#45

Soecka replied:
两者没有类似的 action 镜像，这个仓库只有一个 `pull-request.yml`

TechQuery replied:
> 两者没有类似的 action 镜像，这个仓库只有一个 `pull-request.yml`

4 处步骤一样且可升级：
https://github.com/idea2app/React-MobX-Bootstrap-ts/blob/dfc2a854ba245c1b056a07a5c38468585060c3f1/.github/workflows/main.yml#L14-L33

TechQuery commented on eslint.config.mjs:
```diff
@@ -0,0 +1,72 @@
+            'simple-import-sort/imports': 'error'
+        }
+    },
+    // typescript
```
```suggestion
    // TypeScript
```

TechQuery commented on eslint.config.mjs:
```diff
@@ -0,0 +1,72 @@
+            '@typescript-eslint': tseslint.plugin
+        }
+    },
+    // js
```
```suggestion
    // JavaScript
```

TechQuery commented on eslint.config.mjs:
```diff
@@ -0,0 +1,72 @@
+            'consistent-return': 'error'
+        }
+    },
+    // react
```
```suggestion
    // React
```

TechQuery commented on eslint.config.mjs:
```diff
@@ -0,0 +1,72 @@
+        language: 'json/json',
+        plugins: { json }
+    },
+    // markdown
```
```suggestion
    // MarkDown
```

TechQuery commented on package.json:
```diff
@@ -66,7 +75,10 @@
     },
     "lint-staged": {
-        "*.{html,md,less,json,yml,js,ts,tsx}": "prettier --write"
+        "*.{html,md,less,json,yml,js,ts,tsx}": [
```
```suggestion
        "*.{html,md,less,json,yml,js,mjs,ts,tsx}": [
```
`tsconfig.json` 里也开一下 `checkJs`，并 
`include` `*.mjs`。

Soecka committed:
[add] code spell check extension
[polish] comment message

TechQuery commented on package.json:
```diff
@@ -5,52 +5,62 @@
         "editorjs-html": "^3.4.3",
         "idea-react": "^2.0.0-rc.2",
-        "koajax": "^1.1.2",
+        "koajax": "^3.0.0",
```
为啥降了两个小版本？

Soecka replied:
之前 3.0.2 安装不了，不知道是不是 npm 临时炸了

TechQuery commented on package.json:
```diff
@@ -5,52 +5,62 @@
         "lodash": "^4.17.21",
-        "mobx": "^6.12.4",
+        "mobx": "^6.13.1",
         "mobx-github": "^0.3.2",
         "mobx-i18n": "^0.5.0",
         "mobx-react": "^9.1.1",
         "mobx-restful": "^1.0.0",
         "mobx-restful-table": "^2.0.0-rc.1",
```
还有版本可以升，用 Version lens 插件统一清查一下。

Soecka committed:
[update] deploy CI

TechQuery commented on .github/workflows/main.yml:
```diff
@@ -5,37 +5,16 @@ on:
-              if: ${{ !env.VERCEL_TOKEN || !env.VERCEL_ORG_ID || !env.VERCEL_PROJECT_ID }}
-              with:
-                  publish_dir: ./dist
-                  personal_token: ${{ secrets.GITHUB_TOKEN }}
-                  force_orphan: true
+            - uses: actions/checkout@v4
+              if: ${{ env.VERCEL_TOKEN && env.VERCEL_ORG_ID && env.VERCEL_PROJECT_ID }}
```
我说的是让你参考其它模板仓库来把每一个步骤引用的 action 版本升上去，不是让你删减一大堆步骤……

看这里吧：wuhan2020/WebApp#95

Soecka committed:
[update] CI/CD

TechQuery commented on .github/workflows/main.yml:
```diff
@@ -5,33 +5,28 @@ on:
 jobs:
     Build-and-Deploy:
-        runs-on: ubuntu-latest
         env:
             VERCEL_TOKEN: ${{ secrets.VERCEL_TOKEN }}
             VERCEL_ORG_ID: ${{ secrets.VERCEL_ORG_ID }}
             VERCEL_PROJECT_ID: ${{ secrets.VERCEL_PROJECT_ID }}
+        runs-on: ubuntu-latest
```
```suggestion
        runs-on: ubuntu-latest
        permissions:
            contents: write
        env:
            VERCEL_TOKEN: ${{ secrets.VERCEL_TOKEN }}
            VERCEL_ORG_ID: ${{ secrets.VERCEL_ORG_ID }}
            VERCEL_PROJECT_ID: ${{ secrets.VERCEL_PROJECT_ID }}
```

TechQuery commented on .github/workflows/pull-request.yml:
```diff
@@ -5,23 +5,31 @@ on:
-        runs-on: ubuntu-latest
-        permissions:
-            contents: write
         env:
             VERCEL_TOKEN: ${{ secrets.VERCEL_TOKEN }}
             VERCEL_ORG_ID: ${{ secrets.VERCEL_ORG_ID }}
             VERCEL_PROJECT_ID: ${{ secrets.VERCEL_PROJECT_ID }}
+        runs-on: ubuntu-latest
```
```suggestion
        runs-on: ubuntu-latest
        env:
            VERCEL_TOKEN: ${{ secrets.VERCEL_TOKEN }}
            VERCEL_ORG_ID: ${{ secrets.VERCEL_ORG_ID }}
            VERCEL_PROJECT_ID: ${{ secrets.VERCEL_PROJECT_ID }}
```

TechQuery committed:
Update .github/workflows/main.yml

TechQuery committed:
Update .github/workflows/pull-request.yml

Soecka committed:
[add] inject typescript config to eslint

TechQuery commented on eslint.config.mjs:
```diff
@@ -1,49 +1,50 @@
-        files: ['**/*.{ts,tsx}'],
         plugins: {
-            '@typescript-eslint': typescriptPlugin.plugin
+            ['@typescript-eslint']: tsEslint.plugin,
```
```suggestion
            '@typescript-eslint': tsEslint.plugin,
```
下同

TechQuery commented on eslint.config.mjs:
```diff
@@ -1,49 +1,50 @@
-        files: ['**/*.{js,ts,jsx,tsx}'],
-        plugins: { react },
+        languageOptions: {
+            globals: {
+                ...globals.es2020,
+                ...globals.browser
+            },
```
Node.js 呢？

Soecka committed:
[add] eslint config global variable
[polish] import sort for some component
[add] title for TSXSample
[fix] form event for react

TechQuery commented on package.json:
```diff
@@ -5,68 +5,77 @@
-        "react-bootstrap": "^2.10.3",
+        "mobx-restful": "^1.0.1",
+        "mobx-restful-table": "^2.0.0",
+        "react": "18.3.1",
+        "react-dom": "18.3.1",
```
```suggestion
        "react": "^18.3.1",
        "react-dom": "^18.3.1",
```
前端库一般不需要锁版本，你需要重新执行一下安装命令更新锁文件。

TechQuery commented on package.json:
```diff
@@ -5,68 +5,77 @@
-        "@types/node": "^18.19.39",
+        "@types/lodash": "^4.17.7",
+        "@types/node": "^20.16.5",
+        "@types/react": "18.3.5",
+        "@types/react-dom": "18.3.0",
```
```suggestion
        "@types/react": "^18.3.5",
        "@types/react-dom": "^18.3.0",
```

TechQuery commented on package.json:
```diff
@@ -5,68 +5,77 @@
         "buffer": "^6.0.3",
-        "husky": "^9.0.11",
-        "lint-staged": "^15.2.7",
+        "eslint": "9.10.0",
```
```suggestion
        "eslint": "^9.10.0",
```

TechQuery commented on src/index.tsx:
```diff
@@ -35,4 +35,6 @@ serviceWorker?.addEventListener('controllerchange', () =>
 );
 
-render(<PageFrame />, document.querySelector('#app'));
+const root = createRoot(document.getElementById('app'));
+
+root.render(<PageFrame />);
```
```suggestion
createRoot(document.querySelector('#app')).render(<PageFrame />);
```

Soecka committed:
[polish] dependencies version

Soecka committed:
[polish] ReadMe

TechQuery commented on ReadMe.md:
```diff
@@ -12,7 +12,7 @@
 
 -   Language: [TypeScript v5][2]
--   Component engine: [PReact v10][9]
+-   Component engine: [React v18][9]
```
```suggestion
-   Component engine: [React v18][1]
```
直接复用现有链接即可

TechQuery commented on ReadMe.md:
```diff
@@ -72,7 +72,7 @@ pnpm build
 [7]: https://codespaces.new/idea2app/React-MobX-Bootstrap-ts
 [8]: https://gitpod.io/?autostart=true#https://github.com/idea2app/React-MobX-Bootstrap-ts
-[9]: https://preactjs.com/
+[9]: https://react.dev/
```
```suggestion
```

TechQuery committed:
Update ReadMe.md

TechQuery committed:
Update ReadMe.md
